### ISSUE TITLE
Write information "How many users have voted" into redis

### ISSUE THREAD

**Issue by ostcar** [OpenSlides/openslides-vote-service#16]

**Comment by jsangmeister** [comment]
Mmh, I think it would require some changes in the client, since currently it is just another field in the model, while with this suggestion the client would have to call an URL to get the information. Also, it is a lot slower to send a new request for this information. I think the better solution is keeping the functionality in the AU service.

**Comment by ostcar** [comment]
I talked with @emanuelschuetze about this.

The information how many users have voted can be saved in the datastore. So the vote-service could save this regular (for example once a second). Then this information could be requested via the autoupdate-service. If I should implement this, I would need a internal backend action where to save this information.

emanuel says, that there was no "have i voted" url in os3 and he thinks, that it might not be necessary in os4. So we could skip this for the moment.

**Comment by jsangmeister** [comment]
> emanuel says, that there was no "have i voted" url in os3 and he thinks, that it might not be necessary in os4. So we could skip this for the moment.

There was no such URL because the backend calculated this information from the database. The client needs this information so he can display the poll accordingly - if the user has not voted, there will be the "normal" poll with options to click, while after voting, there only is a checkmark, so you can't vote again. If we don't deliver this information in any way to the client, he must always present the poll with all options, which would be very confusing to the user.

**Comment by jsangmeister** [comment]
Just talked with @emanuelschuetze and @GabrielInTheWorld and this has higher priority than #57. The solution we came up with to prevent polling of any kind and have real-time information in the client is to use the same redis stream as the datastore in the vote service and the vote service just acts as if he was the datastore and writes the information for exactly this fqfield into the redis stream as an update. It is crucial that the field set which the vote service and the datastore write stay disjunct so no information is lost/overwritten. In this case, it is just one field (which should not exist in the models file so it never gets written by the datastore), so it should be no problem.

The autoupdate service then restricts and delivers the field to all clients (i.e. the ones allowed to see the field). I think this is a pretty neat solution which requires no additional routes or polling or similar. What do you think @ostcar?

**Comment by ostcar** [comment]
This could work but I would prefer to add a field to the poll collection in the models.yml.

One reason is, that I parse the models.yml to get the restriction mode. So the autoupdate-service can only deliver fields that are defined in the models.yml. So at least, we have to add the field to the models.yml as calculated field like [projection/content](https://github.com/OpenSlides/OpenSlides/blob/e9d0c07eb49cbdd75757ace2dde8e5df3dd32e0f/docs/models.yml#L3061-L3064).

Another problem is, that the autoupdate-service does not only use the message-bus but also sends requests to the datastore-reader. The autoupdate-service only uses the message-bus to update the cache. So updates from the message-bus for fields that are not in the cache are ignored. A field is added to the cache, when the client first requests it. Then the autoupdate-service asks the datastore-reader for that field.

If the field does not exist in the datastore-reader, it would return `null`. So the autoupdate-service would send to the client, that the field does not exist. Only when there is an update to the field after the request to the autoupdate-service was sent (the time of the responce from the datastore-reader is irrelevant) then the message-bus is used.

If there are regular updates to the field, then this might not be a problem. But it could be a problem. For example if the autoupdate-service gets restarted (what until now is not a problem) or the cache gets cleared (currently not implemented but might be in the future).

In conclusion: We could do it like this, but I think its a hacky solution that could result in strange behaviour or problems in the future.

Maybe we could talk via telephone about this?

**Comment by jsangmeister** [comment]
Mmh, ok. I don't think this is a big problem, because typically the polls will be short and the AU service will not be restarted during a poll. So the client would request the field -> AU returns `null` (which is more or less correct, since no one voted yet) -> vote service writes first update to redis -> from now on, the client receives the correct result. But you are correct that this is not optimal and might lead to problems especially with long polls. 

Could you alternatively implement some kind of custom fetch functions for specific fields, so that this field is not read from the datastore but instead from the vote service? Of course the vote service would then again have to provide a route for this.

Sure we can also talk via telephone about this.

**Comment by ostcar** [comment]
I already though about a custom fetch method. I could try to use calculated fields for this. I am not sure if it works, but it might be worth a try.

One problem is, that it would add a new dependency from the autoupdate-service to the vote-service. So we have to add the url to the vote-service to the autoupdate service. I don't think this is a big problem, but it is one more step that OpenSlides gets a `distributed big ball of mud`.

**Comment by jsangmeister** [comment]
> I already though about a custom fetch method. I could try to use calculated fields for this. I am not sure if it works, but it might be worth a try.

Ok. Post an update here whether this works or not.

> One problem is, that it would add a new dependency from the autoupdate-service to the vote-service. So we have to add the url to the vote-service to the autoupdate service. I don't think this is a big problem, but it is one more step that OpenSlides gets a `distributed big ball of mud`.

I also don't think this is a problem, since it is a meaningful dependency: The AU service distributes data to the client and the vote service has data which is stored only there, so it makes sense to have this dependency. I think the "big ball of mud" comes more from other places than from adding a dependency like this :D

**Comment by ostcar** [comment]
I created a proof of concept for the autoupdate-service that worked (see [here](https://github.com/OpenSlides/openslides-autoupdate-service/pull/363)).

I added the field `poll/vote_count` to the [restriction wiki](https://github.com/OpenSlides/OpenSlides/wiki/poll-restrictions). Is it correct that only poll-manager can see them in the started state?

I added the field in the models.yml. This is required for the restricter: https://github.com/OpenSlides/openslides-backend/pull/1022

The next step is to change the vote-service to save the changes on the message-bus and handle get requstes from the autoupdate-service.

As last step, the proof-of-concept has to be finished to accept environment variables to the vote-service and remove all the duplicated code.

**Comment by jsangmeister** [comment]
The restrictions seem correct. I don't know why the LOS manager needs to see the field, but this was probably also the case before?

**Comment by ostcar** [comment]
> The restrictions seem correct. I don't know why the LOS manager needs to see the field, but this was probably also the case before?

This is a good question. I have no idea. It sounds strange. Maybe @emanuelschuetze or @normanjaeckel know this?

**Comment by peb-adr** [comment]
When I test the setup and submit a vote the following error occurs:

    Saving vote count 1 failed: saving vote_count of poll 22 to message bus: publish poll/22/vote_count, [49] in redis: ERR wrong number of arg
uments for 'xadd' command

The vote is still registered and shown when closing the poll, but the love count is not updated.

Not sure if this know or not, but this should be the relevant issue right?

**Comment by ostcar** [comment]
@peb-adr I found a stupid bug in the autoupdate repo (it tried to access vote9013 instead of vote:9013) and in the vote-service (the vote_count handler was not registered, so 404 was redtuend).

I hope it works now.

**Comment by ostcar** [comment]
> The vote is still registered and shown when closing the poll, but the love count is not updated.

I like the "love count" ;)

> Not sure if this know or not, but this should be the relevant issue right?
After I fixed the bug in the autoupdate-service and vote-serive, the counter is still not updated. But I think this is a bug in the client. When I directly request the field from the autoupdate-service, I get the updates.

Note that this can currently only work for the superadmin. For any other user the field vote_count gets restricted until [this](https://github.com/OpenSlides/openslides-backend/pull/1034) is merged and the autoupdate-service updated.

**Comment by peb-adr** [comment]
> > The vote is still registered and shown when closing the poll, but the love count is not updated.
> 
> I like the "love count" ;)

Oh well, guess I was ready for my weekend :P

**Comment by ostcar** [comment]
Since this works on the server, I opened an issue in the client repo: https://github.com/OpenSlides/openslides-client/issues/674
